Pushover glances

efidoman opened the pull request:
Pushover has a new powerful glances API that sends complications to an apple watch or small bits of data to a smart watch face. More information at the following web page: https://pushover.net/api/glances

This change enhances the pushover notification script to optionally send information to the pushover glances API for watch face updates once every 10 minutes. It works well for apple watch complications. It has not yet been tested for android wear watchface updates yet.

If the openaps parameter, pushoverGlances does not exist in preferences.json or is set to false, then nothing is done or sent related to pushover glances. If pushoverGlances is set to true in preferences.json, then this code will fire every 10 minutes to update pushover glances / smart watch complications.

Note that apple limits the daily total number of complication updates to save watch battery, so we cannot update this glance data more often than every 10 minutes today.

Apple watch face screenshots below (large watchface and Modular)
![img_5637](https://user-images.githubusercontent.com/26698840/33631956-265fb7f2-d9da-11e7-8493-ee1e7f340c76.JPG)
![img_5639](https://user-images.githubusercontent.com/26698840/33631957-266f7a84-d9da-11e7-9ed3-84cdbf795741.JPG)
![img_5642](https://user-images.githubusercontent.com/26698840/33631958-2692daf6-d9da-11e7-9390-5ed0ee93bb26.JPG)
![img_5645](https://user-images.githubusercontent.com/26698840/33631959-26afa28a-d9da-11e7-8f56-0db7ab393ff8.JPG)
![img_5647](https://user-images.githubusercontent.com/26698840/33631960-26bfe528-d9da-11e7-8107-6662cbbc4efa.JPG)
![img_5648](https://user-images.githubusercontent.com/26698840/33631961-26cf00ee-d9da-11e7-9bab-d71286510d5f.JPG)
![img_5649](https://user-images.githubusercontent.com/26698840/33631962-26e6b61c-d9da-11e7-89d8-2989acb46bca.JPG)
![img_5651](https://user-images.githubusercontent.com/26698840/33631963-26fb0036-d9da-11e7-92fd-f3d44ae2e72b.JPG)
![img_5655](https://user-images.githubusercontent.com/26698840/33631964-2715b340-d9da-11e7-9bb2-8acabbb9c6b2.JPG)
![img_5676](https://user-images.githubusercontent.com/26698840/33631965-2722f582-d9da-11e7-839b-32b42e257c13.JPG)
![img_5677](https://user-images.githubusercontent.com/26698840/33631966-273998e6-d9da-11e7-95f6-48f522f05322.JPG)

scottleibrand commented on bin/oref0-pushover.sh:
```diff
@@ -28,3 +28,76 @@ elif [[ $ONLYFOR =~ "insulin" ]] && ! cat $FILE | egrep "maxBolus"; then
+GLUCOSE="monitor/glucose.json"
+if [ ! -f $GLANCES ]; then
+  # First time through it will get created older than 10 minutes so it'll fire
+  touch $GLANCES && touch -r $GLANCES -d '-11 mins' $GLANCES
```
I think your first `touch` here is redundant.

efidoman replied:
I'll remove that first touch. Also, I'm currently refactoring it to perform faster by reading values from smb-suggested.json only. Give me a day or two to make these changes and test it.

efidoman commented:
@scottleibrand - I finished the refactoring, but at the same time I added another notification option for ifttt. Let me know if this needs to go somewhere else but I felt it fit best here so I added another optional notification that will send an ifttt maker event "carbs-required" if additional carbs are required and if environment variable "MAKER_KEY" is set. This is the ifttt webhooks Maker key https://ifttt.com/maker_webhooks

The "carbs-required" event can then be used with a ifttt recipe to perform an action based on carbs required. A good use case is to use the ifttt phone action to get a phone call with this event that will read out in human language the additional carbs and other vital facts. It will leave a voice mail if not answered.

scottleibrand commented on bin/oref0-pushover.sh:
```diff
@@ -28,3 +28,94 @@ elif [[ $ONLYFOR =~ "insulin" ]] && ! cat $FILE | egrep "maxBolus"; then
+      direction="+"
+    else
+      direction="++"
+    fi
```
Normally direction arrows are based on something like a 15m average.  Do we have the avgDelta available to use here?

efidoman replied:
Is there an avgDelta stored somewhere in an openaps json file? I could calculate it from glucose.json if not.

scottleibrand replied:
No, we calculate avgDelta in determine-basal, but don't write it out anywhere.  I guess we could either add that to the enacted json, or recalculate it here.  Or you could do something more like what @cjo20 did with the mdt-trend-arrow stuff...

efidoman replied:
I looked at the mdt-trend-arrow code and it definitely looks like something that xdrip-js-logger should be doing. If it is done there, then any downstream consumer of xdrip-js-logger glucose data would have better trend arrows including NS. Let me look at that in the next couple of days. Then this apple watch pushover glances update could just use the trend arrow from the glucose data directly. @scottleibrand - What do you think?

scottleibrand commented on bin/oref0-pushover.sh:
```diff
@@ -28,3 +28,94 @@ elif [[ $ONLYFOR =~ "insulin" ]] && ! cat $FILE | egrep "maxBolus"; then
+    touch $GLANCES && touch -r $GLANCES -d '-11 mins' $GLANCES
+  fi
+
+  if test `find $GLANCES -mmin +10`
```
If you have more than one rig, this will be too often.  Do we need to run it less often in that scenario to avoid burning through the complications budget for the day? Or will the watch just ignore the ones sent more than once every 10 minutes?  I wonder if this is something that we should eventually port over to NS to avoid duplicate notifications from multiple rigs.

efidoman replied:
Good point about multiple rigs. We really only want one rig to do this work ... the one that is currently active. I would need to test it, but I think we are still safe in this scenario because the non-active rigs would have an enact time > 5 minutes right? In this case the entire script exits at the beginning.

scottleibrand replied:
When running multiple rigs in the same or adjacent rooms, it's very common for one rig to enact something, and then as soon as it's done another rig reads the pump and enacts something slightly different.

scottleibrand commented:
Yeah, I prefer the idea of using the CGM trend arrows. That also makes it possible to ensure the arrows always match the normal Dexcom receiver/app for everyone using that.

efidoman commented:
Ok - I'll refactor this now to use the glucose trend arrows. That way it shows whatever the CGM or collector-system put there.

efidoman commented:
@scottleibrand - I refactored it to use the direction field in the latest glucose.json entry. I tested it and it works good for me. look at the latest commit. I had some weird cut-n-paste stuff going on with the first 7546b55 commit.

efidoman commented:
I figured out why my carbs-required phone calls stopped working since 12/14. It seems ifttt limits the phone calls to 100 per month without documenting that limit. I did just find the email they sent me though 
---
Hi feibelman,You have reached IFTTT's limit of 100 phone calls.All of your active Applets that make phone calls are paused until Monday, January 01, 2018.
---

I was doing a lot of testing at the time and causing carbs-required events to occur by giving manual boluses and delaying carb entry. However, 100 means an average of 3 per day. I believe that I should tighten up the parameters a bit for a carbs-required maker event and avoid repeat notifications if possible. Not everyone will use this to trigger the phone action, so maybe these need to be parameters. I think the # of carbs-required before notification is already a parameter (I will check).

sulkaharo commented:
FYI I asked pushover and they’re apparently not kidding about the 1 message / 20 minutes limitation when using this API on iOS, so any functionality should be planned accordingly.

efidoman commented:
@sulkaharo - Thanks for the heads up. I'm OK with considering a change to the default pushover glance to once every 20 minutes. However, I've been testing the pushover glances notification to my apple watch with 10 minute updates for several weeks now and I get all the updates. It doesn't seem to throttle me at the 10 minute rate.

Earlier in the testing cycle, when I tried 5 minute updates, I reached my limit sometimes around mid-day, so it seems 10 minutes may be the real quota. I understand Apple setting limits, but vague limits that aren't clearly understood are a little frustrating. ;->

efidoman commented:
I've been running this for the last 3 full weeks using the pushover glances at 10 minute intervals and it hasn't missed an update or ran out of quota for the day. I'd say the 10 minute interval is most likely OK for this use case.

scottleibrand commented:
Cool. Do we need any other tweaks for multi-rig support, or are we good to merge?

efidoman commented:
Is there an quick solution to support multiple rigs or would it require porting it to work in openaps loop?

If it requires porting it to run in the loop, can you give me an overview of what needs to be done and in which files?
